complete.py

```diff
@@ -0,0 +1,795 @@
+        if self.platform == 'cpu':
+            torch.set_num_threads(2)
+        else:
+            raise Exception(f"we don't support gpu just yet")
+
+        self.species = self.model.species_to_tensor(atoms).to(self.device).unsqueeze(0)
+        self.temperature = temperature
+        self.beta = 1.0 / (kB * temperature)
+
+        self.W_shads = []
+        self.W = []
+
+    def minimize(self,
+                 coords: unit.quantity.Quantity,
+                 maxiter: int = 1000):
+        """
+        Minimizes the molecule. Note, we usually don't want to do this given an input structure since they are already distributed i.i.d.
+
+        Parameters
+        ----------
+        coords:simtk.unit.quantity.Quantity
+        maxiter: int, default:1000
+            Maximum number of minimization steps performed.
+
+        Returns
+        -------
+        coords:simtk.unit.quantity.Quantity
+        """
+
+        assert (type(coords) == unit.quantity.Quantity)
+
+        x = coords.value_in_unit(unit.angstrom)
+        self.memory_of_energy = []
+        _logger.info("Begin minimizing...")
+        f = optimize.minimize(self._target_energy_function, x, method='BFGS',
+                              jac=True, options={'maxiter': maxiter, 'disp': True})
+
+        _logger.critical(f"Minimization status: {f.success}")
+        memory_of_energy = copy.deepcopy(self.memory_of_energy)
+        self.memory_of_energy = []
+
+        return f.x.reshape(-1, 3) * unit.angstrom, memory_of_energy
+
+    def calculate_force(self,
+                        x: unit.quantity.Quantity) -> (unit.quantity.Quantity, unit.quantity.Quantity):
+        """
+        Given a coordinate set the forces with respect to the coordinates are calculated.
+
+        Parameters
+        ----------
+        x : array of floats, unit'd (distance unit)
+            initial configuration
+
+        Returns
+        -------
+        F : float, unit'd
+        E : float, unit'd
+        """
+        #assert (type(x) == unit.quantity.Quantity)
+
+        coordinates = torch.tensor([x.value_in_unit(unit.angstroms)],
+                                   requires_grad=True, device=self.device, dtype=torch.float32)
+
+        energy_in_hartree = self._reform_as_energy_tensor(coordinates)
+
+        # derivative of E (in kJ/mol) w.r.t. coordinates (in nm)
+        derivative = torch.autograd.grad((energy_in_hartree).sum(), coordinates)[0]
+
+        if self.platform == 'cpu':
+            F = -1 * derivative[0].numpy()
+        elif self.platform == 'cuda':
+            F = - np.array(derivative.cpu())[0]
+        else:
+            raise RuntimeError('Platform needs to be specified. Either CPU or CUDA.')
+
+        return (F * self.hartree_to_kJ_per_mole * (unit.kilojoule_per_mole / unit.angstrom),
+                energy_in_hartree.item() * self.hartree_to_kJ_per_mole * unit.kilojoule_per_mole)
+
+    def _reform_as_energy_tensor(self, coordinates: torch.tensor):
```

Comment:
at some point this function could be deleted and simply replaced in code with 
`energy_in_hartree = self.model((self.species, coordinates)).energies`